ypbind does not bind to ypserv on localhost (same machine) if NetworkManager
thinks that the network is off.

rhbz#829487

diff -up ypbind-mt-1.20.4/src/local.h.nmlocal ypbind-mt-1.20.4/src/local.h
--- ypbind-mt-1.20.4/src/local.h.nmlocal	2006-09-13 14:53:43.000000000 +0200
+++ ypbind-mt-1.20.4/src/local.h	2012-07-10 10:34:38.932528706 +0200
@@ -7,6 +7,7 @@ extern int broken_server;
 extern int port;
 extern int ping_interval;
 extern int use_broadcast;
+extern int localhost_used;
 extern char *domain;
 
 extern void find_domain (const char *domain, ypbind_resp *result);
diff -up ypbind-mt-1.20.4/src/serv_list.c.nmlocal ypbind-mt-1.20.4/src/serv_list.c
--- ypbind-mt-1.20.4/src/serv_list.c.nmlocal	2012-07-10 10:34:38.926528667 +0200
+++ ypbind-mt-1.20.4/src/serv_list.c	2012-07-10 10:34:38.933528712 +0200
@@ -108,6 +108,28 @@ static pthread_mutex_t search_lock = PTH
 static void do_broadcast (struct binding *list);
 static int ping_all (struct binding *list);
 
+/* We have localhost defined in one of the domains.
+ * If so, we don't need to be connected to outer network. */
+void
+check_localhost()
+{
+  int i, s;
+  localhost_used = 0;
+  for (i = 0; i < max_domains; ++i)
+    {
+      for (s = 0; s < _MAXSERVER; ++s)
+        {
+	  if (domainlist[i].server[s].host == NULL)
+	    break;
+          if (strncmp(inet_ntoa(domainlist[i].server[s].addr), "127", 3) == 0)
+            {
+       	      localhost_used = 1;
+      	      return;
+            }
+        }
+    }
+}
+
 static void
 remove_bindingfile (const char *domain_name)
 {
@@ -216,6 +238,7 @@ update_bindingfile (struct binding *entr
     }
   else
     log_msg (LOG_ERR, "open(%s): %s", path2, strerror (errno));
+  check_localhost();
 }
 
 /* this is called from the RPC thread (ypset). */
@@ -568,6 +591,7 @@ add_server (const char *domain, const ch
 	 If there is none, use the first one. */
       memcpy (&entry->server[active].addr, hent->h_addr_list[0],
 	      hent->h_length);
+      check_localhost();
       res = 1;
     }
 
@@ -1086,7 +1110,7 @@ test_bindings (void *param __attribute__
   int lastcheck = 0;
 
 #ifdef USE_DBUS_NM
-  if (is_online)
+  if (is_online || localhost_used)
 #endif
     do_binding ();
 
@@ -1106,7 +1130,7 @@ test_bindings (void *param __attribute__
 	lastcheck = 0;
 
 #if USE_DBUS_NM
-      if (is_online)
+      if (is_online || localhost_used)
 
 #endif
 	lastcheck = test_bindings_once (lastcheck, NULL);
diff -up ypbind-mt-1.20.4/src/ypbind_dbus_nm.c.nmlocal ypbind-mt-1.20.4/src/ypbind_dbus_nm.c
--- ypbind-mt-1.20.4/src/ypbind_dbus_nm.c.nmlocal	2012-07-10 10:34:38.929528686 +0200
+++ ypbind-mt-1.20.4/src/ypbind_dbus_nm.c	2012-07-10 10:34:38.935528725 +0200
@@ -86,8 +86,11 @@ go_offline (void)
   if (debug_flag)
     log_msg (LOG_DEBUG, _("Switch to offline mode"));
   is_online = 0;
-  portmapper_disconnect ();
-  clear_server ();
+  if (!localhost_used)
+    {
+      portmapper_disconnect ();
+      clear_server ();
+    }
 }
 
 static void
diff -up ypbind-mt-1.20.4/src/ypbind-mt.c.nmlocal ypbind-mt-1.20.4/src/ypbind-mt.c
--- ypbind-mt-1.20.4/src/ypbind-mt.c.nmlocal	2012-07-10 10:34:38.916528601 +0200
+++ ypbind-mt-1.20.4/src/ypbind-mt.c	2012-07-10 10:34:38.936528732 +0200
@@ -73,6 +73,7 @@ int use_broadcast = 0;
 int broken_server = 0;
 int ping_interval = 20;
 int local_only = 0;
+int localhost_used = 1;
 int port = -1;
 static int lock_fd;
 static int pid_is_written = 0;
@@ -902,7 +903,7 @@ main (int argc, char **argv)
     }
 
 #ifdef USE_DBUS_NM
-  if (!is_online)
+  if (!is_online && !localhost_used)
     portmapper_disconnect ();
 #endif
 
